### PR TITLE
[WIP] vulkan compute

### PR DESCRIPTION
**Pull request by nihui**
- [x] find vulkan library
- [x] shader compile
- [x] device management
- [x] blob memory allocator
- [x] weight memory allocator
- [x] workspace memory allocator
- [x] staging buffer memory allocator
- [x] vkmat
- [x] new layer interface
- [x] descriptor update
- [x] dynamic workgroup
- [x] the command class
- [x] low level op api
- [x] op convolution
- [x] op convolutiondepthwise
- [x] op pooling
- [x] op relu
- [x] op innerproduct
- [x] op softmax
- [x] op split
- [x] op concat
- [x] op batchnorm
- [x] op padding
- [x] op scale
- [x] op dropout
- [x] op eltwise
- [ ] op reshape
- [x] op flatten
- [x] op binaryop
- [x] op unaryop
- [ ] high level api
- [x] optional VK_KHR_descriptor_update_template
- [x] optional VK_KHR_push_descriptor

```

int main()
{
    srand(time(NULL));

    ncnn::create_gpu_instance();

    {

    ncnn::VulkanDevice vkdev;

    {
    // high-level api
    ncnn::Net net;

    net.use_vulkan_compute = 1;
    net.set_vulkan_device(&vkdev);

    net.load_param("mnist.param");
    net.load_model("mnist.bin");

    {

    ncnn::Extractor ex = net.create_extractor();

    ncnn::Mat in(28, 28, 1);
    in.fill(0.5f);
    ex.input("data", in);

    ncnn::Mat ip2;
    ex.extract("ip2", ip2);

    pretty_print(ip2);

    }

    }

    }

    ncnn::destroy_gpu_instance();

    return 0;
}

```

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=618) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=618) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/ncnn?pullRequest=618) it.</sub>

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=618) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=618) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/ncnn?pullRequest=618) it.</sub>

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=618) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=618) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/ncnn?pullRequest=618) it.</sub>

**Comment by nihui** [comment]
TODO
forward 的时候需要加 state checking，避免重复加 barrier
vktransfer 应当放在 transfer queue 上执行，有硬件dma加速
大size要遵循 VK_KHR_dedicated_allocation
集成显卡有 unified 内存去掉staging buffer减少copy
blob fifo 规则可以用 ringbuffer 做
workspace 即用即销毁可以只用一个大size
staging 输入输出用一下简单点做个块分配器
weight 预先可以知道多大不考虑释放最简单

**Comment by hgj3804278** [comment]
按照你这个版本，启用vulkan开启的Gpu计算结果和cpu结算结果不一样，cpu计算结果正确，而gpu计算结果不正确

**Comment by hgj3804278** [comment]
按照你这个版本，启用vulkan开启的Gpu计算结果和cpu结算结果不一样，cpu计算结果正确，而gpu计算结果不正确
